cub/cub/device/dispatch/dispatch_merge.cuh

```diff
@@ -218,13 +218,8 @@ template <typename KeyIt1,
           typename PolicyHub = device_merge_policy_hub<value_t<KeyIt1>, value_t<ValueIt1>>>
 struct dispatch_t
 {
-  using key_t   = cub::detail::value_t<KeyIt1>;
-  using value_t = cub::detail::value_t<ValueIt1>;
-
   // Cannot check output iterators, since they could be discard iterators, which do not have the right value_type
-  static_assert(::cuda::std::is_same<cub::detail::value_t<KeyIt2>, key_t>::value, "");
-  static_assert(::cuda::std::is_same<cub::detail::value_t<ValueIt2>, value_t>::value, "");
```

Comment:
suggestion: @miscco we require this in our [docs](https://nvidia.github.io/cccl/thrust/api/function_group__merging_1ga3d2776685a00dca265399f411784acec.html):

> InputIterator1 and InputIterator2 have the same value_type

 If we don't want to enforce that, we should relax requirements in the docs (both Thrust and CUB) and add tests. I think it's fine to merge this PR without doing that, but we should at least file an issue.

Reply:
So people are already doing it as we have seen with cuDF.

I believe we should think about something in the direction of `is_assignable<value_t<OutIt>&, value_t<InIt1>>` with the obvious caveat that it needs to work for discard iterators.

Reply:
I would argue it should be `is_assignable<reference_t<OutIt>, value_t<InIt1>>`. That would handle proxy references.